.circleci/parse_yaml.py

```diff
@@ -0,0 +1,269 @@
+            sys.exit(1)
+
+
+    def download(self):
+        """Check if provided weblink to the file exists. 
+        Download if verified.
+        """
+        path = self.yaml_data['data_file']['location']['url']
+        self.downloaded_file = path.split('/')[-1]
+
+        outdir_root = 'submitted_data/'
+        pathlib.Path(outdir_root).mkdir(exist_ok=True)
+        self.outdir = outdir_root + str(self.data_type) + '/'
+        pathlib.Path(self.outdir).mkdir(exist_ok=True)
+
+        if verify_weblink(path):
+            logging.info(f' starting download')
+            download_file(path, self.outdir)
+            self.hash = self.yaml_data['data_file']['location']['md5']
+        else:
+            logging.error(f' file unavailable')
+            sys.exit(1)
+
+
+    def bulk_download(self):
+        """Bulk download option assumes one of the following file types:
+        ['tgz', 'tar.gz', 'tar.bz2', 'tar.xz']
+        """
+        import tarfile
+
+        self.download()
+
+        if tarfile.is_tarfile(self.downloaded_file):
+            tarred_data = tarfile.open(
+                self.downloaded_file, 
+                mode='r:*'
+            )
+            tarred_data.extractall(path=self.outdir)
+    
+
+    def check_hash(self):
+        """Compare md5 of the downloaded file to the provided value"""
+        if validate_hash(self.downloaded_file, self.hash):
+            logging.info(f' data is ready for cimr processing.')
+            return True
+        else:
+            logging.error(f' provided md5 hash didn\'t match the downloaded file.')
+            return False
+
+
+    def check_defined(self):
+        import os
+
+        """Check whether the submitted data is a single file"""
+        if self.yaml_data['defined_as'] == 'upload':
+            self.download()
+            self.check_hash()
+        elif self.yaml_data['defined_as'] == 'upload_bulk':
+            self.bulk_download()
+            if self.check_hash():
+                os.remove(self.outdir + self.downloaded_file)
+        else:
+            logging.error(f' accepted \'defined_as\' variables are \'upload\' and \'upload_bulk\'.')
+            sys.exit(1)
+
+
+    def check_data_file(self):
+        """Standard set of Yamler functions to check information on the
+        contributed data file for ci cimr processing.
+        """
+        self.set_data_type()
+        self.check_defined()
+
+
+if __name__ == '__main__':
+
+    if len(sys.argv) == 1:
+        yaml_file = predefine_yaml()
+    else:
+        yaml_file = pathlib.Path(sys.argv[1])
+
+    try:
+        yaml_file_path = yaml_file.resolve(strict=True)
+        logging.info(f' processing metadata {yaml_file_path}.')
+        yaml_data = load_yaml(yaml_file)
+        print(yaml_data)
+        y = Yamler(yaml_data)
+        y.check_data_file()
+    except FileNotFoundError:
+        logging.info(f' no new yaml file found to process.')
+        sys.exit(0)
```

Comment:
Non-zero exit code will be better.

Reply:
I changed this to a zero exit since CI may run without a new yaml to process.

Reply:
That's a good point. Thanks. Is it better to call the `except` block earlier, like this:
```python
    try:
        yaml_file_path = yaml_file.resolve(strict=True)
    except FileNotFoundError:
        logging.info(f' no new yaml file found to process.')
        sys.exit(0)

    logging.info(f' processing metadata {yaml_file_path}.')
    yaml_data = load_yaml(yaml_file)
    print(yaml_data)
    y = Yamler(yaml_data)
    y.check_data_file()
```
This will ensure that any other `FileNotFoundError` exception raised later will terminate the program (and CircleCI process).

Reply:
ah. yes, that makes sense. thanks.